### ISSUE TITLE
just_audio_web wasm support

### ISSUE THREAD

**Issue by renefloor** [ryanheise/just_audio#1127]
**Is your feature request related to a problem? Please describe.**
Not really a problem, was playing around with a hobby project and noticed it's not possible to use dart:html when trying web assembly (wasm). Wasm is highly experimental, so don't expect any fix soon, just documenting the issue.

<img width="858" alt="image" src="https://github.com/ryanheise/just_audio/assets/15101411/a11b77ef-c7d9-4f0b-8189-03694014d4c6">

**Describe the solution you'd like**

I was looking at the PR made in url_launcher_web to get a feeling what kind of change is needed: https://github.com/flutter/packages/pull/5451/files#diff-dcfb51504389f45904bc24488d51b8684fb9bd7e31bc7dbc79ac0b40a020e66e

It looks like we might be able to replace the `dart:html` package with the highly experimental [web](https://pub.dev/packages/web) package.

**Describe alternatives you've considered**
I could look for other audio packages, but as I said, no hurry at all.

**Additional context**
Currently you can't `flutter run wasm`, you can only build, making debugging a bit harder if a migration doesn't succeed in 1 go.

**Comment by ryanheise** [comment]
Thanks for the feature idea. I didn't realise the web package was out of beta already but it appears that it is, which is good news. I had a quick play with it but ran into a bug trying to invoke the audio element constructor. I will report it on the flutter repo when I get home, but if it is indeed a bug in web, then it could be quite some time before just_audio can incorporate it. For example, I also can't currently use the latest version of web because it depends on a newer version of flutter_test than the one pinned in the latest stable SDK.

**Comment by ryanheise** [comment]
I've filed a bug on Flutter: https://github.com/flutter/flutter/issues/139125 .

**Comment by ryanheise** [comment]
https://github.com/dart-lang/web/issues/124

**Comment by ryanheise** [comment]
Implementing a workaround mentioned in the issue linked above, I have created an implementation to test here: #1136 .

**Comment by ryanheise** [comment]
@renefloor Have you had a chance to test #1136 ?

I've tested the implementation on regular web and it works. However, I did not personally do any testing on wasm, since I'm running short on storage and switching back/forth between flutter channels is a bit difficult at the moment.

**Comment by Osmosis311** [comment]
Hi! Any update on this?  Would love to be able to compile to WASM.  Thanks so much!  Amazing plugin!

**Comment by ryanheise** [comment]
As you can see above, I've implemented it, and am just waiting for confirmation from someone whether it works for WASM. If you're willing to test it, please let me know whether it works for you.

**Comment by renefloor** [comment]
I just tried it in an existing project, but got the following error:

```
Target dart2wasm failed: ProcessException: Process exited abnormally with exit code 64:
*NOTE*: Compilation to WasmGC is experimental.
The support may change, or be removed, with no advance notice.


Unhandled exception:
Constant InstanceConstant(const _FfiAbiSpecificMapping{_FfiAbiSpecificMapping.nativeTypes: const <Type?>[Uint32, Uint64, Uint32, Uint64, Uint64, Uint64, Uint64, Uint64, Uint32, Uint64, Uint64, Uint32, Uint64, Uint32, Uint64,
Uint32, Uint64, Uint64, Uint64, Uint64, Uint32, Uint64]}) references field _FfiAbiSpecificMapping.nativeTypes which is not retained
#0      _TreeShakerConstantVisitor.visitInstanceConstant.<anonymous closure> (package:vm/transformations/type_flow/transformer.dart:2257)
#1      _LinkedHashMapMixin.forEach (dart:collection-patch/compact_hash.dart:633)
```

I made the following pubspec change:
```yaml
dependency_overrides:
  just_audio_web:
    git:
      url: https://github.com/ryanheise/just_audio.git
      ref: feature/web_package
      path: just_audio_web
  web: ^0.5.0
```

But I really have no clue what this error is. (So also no clue if it's related to just_audio.) Will try on a clean project another time.

Edit: compilation works. Looks like it was related to a Rive issue, compiled after removing rive. https://github.com/rive-app/rive-flutter/issues/391

For my use case it seems like it works 'just' fine :)
You can test my changes here: (live for 30 days)
https://slide-puzzle-1a294--tst-fhj4jo46.web.app/
compared to older version: https://puzzle.renefloor.nl

Relevant PR: https://github.com/renefloor/flutter_puzzle_hack/pull/5/

I think I'll wait on Rive to actually use wasm, but first impression is good.

**Comment by ryanheise** [comment]
#1136 is now merged. I've published the new just_audio_web on `pub.dev` but not yet bumped the dependency from just_audio to it, so to get this in your project now, you would need to run `flutter pub upgrade`. I'll do the actual bump in just_audio after I merge another PR related to web.

Thanks @renefloor for testing (and also proposing) the feature.

**Comment by daohoangson** [comment]
Any specific reason why we don't support the latest version of the `web` package? Version constraint is currently like this `web: '>=0.3.0 <0.5.0'`

https://github.com/ryanheise/just_audio/blob/b5595f51b19291e46f9e53bb7eadcd0fc647d18b/just_audio_web/pubspec.yaml#L21

**Comment by ryanheise** [comment]
Oh, at the time I did it, I was not in a position to test that because it wasn't stable. I will check this today.

**Comment by renefloor** [comment]
In my test I had to add that to my pubspec overrides, maybe I should have said that more explicit.

**Comment by ryanheise** [comment]
The upper bound has been increased to <0.6.0 in the latest release.

Note: if anyone uses just_audio_background, that has issues which I'm currently working on.

**Comment by ryanheise** [comment]
just_audio_background is now published to sync with the other changes.
